src/crypto.rs

```diff
@@ -23,50 +36,132 @@ pub trait Pseudonymize {
+        let prefix_end = t.iri.rfind(['#', '/']).unwrap();
+        let prefix = &t.iri[0..=prefix_end];
+        let crypted = self.pseudo(t.iri.as_bytes()).to_string();
         return NamedNode {
-            iri: format!("{prefix}{hash}"),
+            iri: format!("{prefix}{crypted}"),
         };
     }
 
+    /// Pseudonymize a literal resulting in a simple literal (hash string).
     fn pseudo_literal(&self, l: &Literal) -> Literal {
         let value = match l {
             Literal::Typed { value, datatype: _ } => value,
             Literal::LanguageTaggedString { value, language: _ } => value,
             Literal::Simple { value } => value,
         };
-        let hash = blake3::hash(value.as_bytes());
-        return Literal::Simple {
-            value: hash.to_string(),
-        };
+        let crypted = self.pseudo(value.as_bytes());
+        return Literal::Simple { value: crypted };
     }
 
+    /// Leave blank nodes unchanged.
     fn pseudo_blank_node(&self, u: &BlankNode) -> BlankNode {
         return u.clone();
     }
 }
+
+/// Available pseudonymization algorithms.
+pub enum Algorithm {
+    Blake3,
+}
+
+impl Default for Algorithm {
+    fn default() -> Self {
+        return Algorithm::Blake3;
+    }
+}
+
+/// Factory method for creating a pseudonymizer
+/// based on the selected algorithm and secret key.
+pub fn new_pseudonymizer(algo: Option<Algorithm>, secret: Option<Vec<u8>>) -> impl Pseudonymize {
+    let pseudonymizer = match algo.unwrap_or_default() {
+        Algorithm::Blake3 => Blake3Hasher::new(secret),
+    };
+
+    return pseudonymizer;
+}
+
+/// BLAKE3-based pseudonymizer.
+pub struct Blake3Hasher {
+    pub key: [u8; 32],
+}
+
+impl Blake3Hasher {
+    pub fn new(secret: Option<Vec<u8>>) -> Self {
+        secret.as_ref().inspect(|s| {
+            if s.len() < 32 {
+                panic!("Secret must be at least 32 bytes long");
+            }
+        });
+
+        // blake3 key must be exactly 32 bytes long
+        let mut key = [0u8; 32];
+        let key_vec = match secret {
+            Some(s) => blake3::hash(&s).as_bytes()[..32].to_vec(),
+            None => generate_key(32),
+        };
+        key.copy_from_slice(&key_vec[..32]);
+
+        return Self { key };
+    }
+}
+
+impl Pseudonymize for Blake3Hasher {
+    fn pseudo(&self, data: &[u8]) -> String {
+        return blake3::keyed_hash(&self.key, data).to_string();
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn is_valid_hex(input: &str) -> bool {
+        input.chars().all(|c| c.is_digit(16))
+    }
+
+    // Test the generation of a cryptographic key
+    #[test]
+    fn test_generate_key() {
+        let key = generate_key(42);
+        assert_eq!(key.len(), 42);
+    }
+
+    #[test]
+    fn test_pseudo_named_node() {
+        let hasher = Blake3Hasher::new(None);
+        let named_node = NamedNode {
+            iri: "http://example.com/abc".to_string(),
```

Comment:
`abc` is valid hex, so if we `pseudo_named_node` returns `named_node` without modifying it the test still passes :stuck_out_tongue: 
I suggest we change it to some non-hex string.
```suggestion
            iri: "http://example.com/tripsu".to_string(),
```

Reply:
nice catch! :sunglasses: